### PR TITLE
Rewrite Kettle class to use GameController

### PR DESCRIPTION
**Pull request by djkaty**
This is a significant re-write of kettle's behaviour to support an AI opponent.

The Kettle class now uses GameController to manage games. This allows an AI opponent to perform turns which then get fed back into Kettle for onward transmission to the game connection, while a human player inputs actions by sending commands over the game connection.

Kettle maintains a single thread per session by integrating with GameController's event loop such that its callbacks are executed on the same thread. This completely obviates the need for thread-safe code or synchronization.

Kettle used an infinite wait loop to read packets before acting. This is not desired behaviour when an AI is acting because the blocking socket receive prevents the AI from performing its actions. Socket receives are now non-blocking (we still have to allow incoming data on the AI's turn in case the human player concedes). If no data is received after a small period, control is relinquished to GameController's event loop so that the AI can act if it is the AI's turn.

Additionally, the await mechanism is used to make Kettle's networking I/O largely asynchronous.

Since socket reads no longer block, we do not want to send an updated game state and options continually. Game state data is now only sent when changes are queued, and options are only sent if a human player is next to act, since we do not want to send option sets for AI players, only game state updates.

Similarly, options and entity choices for human players will only be sent once until a new option choice is received to avoid flooding the connection with repeated data. The KettleManager.options_sent flag controls this.

Finally, a number of unnecessary calls to manager.refresh_full_state() were removed.

NOTE: Due to the design of GameController, if the match is between two human players, no code changes are needed.

NOTE: This version of kettle does not support mulligan behaviour correctly. This will be addressed in a forthcoming PR.

NOTE: This version of kettle always assumes that the 2nd player is an AI. To resolve this and select a human player or a specific AI model, additional protocols are needed in the game connection. This will be addressed in a forthcoming PR.

Tested and working with the HeroPowerAI in the Hearthstone client with stove.

### CODE REVIEW

File: kettle/kettle.py [hunk 1]
```diff
@@ -14,6 +16,8 @@
 from fireplace.game import BaseGame as Game
 from fireplace.player import Player
 from fireplace.utils import CardList
+from fireplace.controller import GameController
+from fireplace.ai.player import PlayerAI
```
Comment (jleclanche) [inline kettle/kettle.py]: Watch out for import ordering

**Comment by djkaty** [comment]
options_sent re-written to be reactive

File: kettle/kettle.py [hunk 2]
```diff
@@ -57,6 +61,8 @@ def action_end(self, type, source):
 	def game_step(self, step, next_step):
 		DEBUG("Game.STEP changes to %r (next step is %r)", step, next_step)
 		self.refresh_full_state()
+		if (step == Step.MAIN_ACTION):
```
Comment (jleclanche) [inline kettle/kettle.py]: No parentheses around if condititional.

File: kettle/kettle.py [hunk 3]
```diff
@@ -148,6 +154,8 @@ def refresh_choices(self):
 		self.queued_data.append(payload)
 
 	def refresh_options(self):
+		if isinstance(self.game.current_player, PlayerAI):
+			return
```
Comment (jleclanche) [inline kettle/kettle.py]: Wait what? Why?

Comment (djkaty) [inline kettle/kettle.py]: Because we don't send AI options to a human client.

Comment (jleclanche) [inline kettle/kettle.py]: You got that mixed up. Also, there's no reason not to send options - AI clients may want them.

Comment (djkaty) [inline kettle/kettle.py]: There are no AI clients on the wire, they run locally in fireplace under the GameController. refresh_options() only queues to send to stove.

Comment (djkaty) [inline kettle/kettle.py]: You do realize I tested this before PRing it right? :)

**Comment by shinoi2** [comment]
Too old and has conflict, close it.

**Pull request closed without merging**